src/macros.jl

```diff
@@ -1005,34 +1005,151 @@ function parsevariable(_error::Function, infoexpr::VariableInfoExpr, lvalue, lsi
-#   `buildvariable` since they may be altered by the parsing of `expr` and we may need to pass it pointwise if it is a container since
-#   `buildvariable` is called separately for each variable of the container. Moreover they will be passed inside the VariableInfo object.
-# * A custom error function is passed as positional argument to print the full @variable call before the error message.
-# * The keyword argument basename is not passed to `constructconstraint!` but rather to `addvariable`.
-#
-# Examples:
-# * `@variable(m, x >= 0)` is equivalent to `x = addvariable(m, buildvariable(msg -> error("In @variable(m, x >= 0): ", msg), VariableInfo(true, 0, false, NaN, false, NaN, false, NaN, false, false)), "x")`
-# * `@variable(m, x[1:N,1:N], Symmetric, Poly(X))` is equivalent to
-#   ```
-#   x = Matrix{...}(N, N)
-#   for i in 1:N
-#       for j in 1:N
-#           x[i,j] = x[j,i] = addvariable(m, buildvariable(msg -> error("In @variable(m, x[1:N,1:N], Symmetric, Poly(X)): ", msg), VariableInfo(false, NaN, false, NaN, false, NaN, false, NaN, false, false), Poly(X)), "")
-#       end
-#   end
-#   ```
+"""
+    @variable(m; kwargs...)
+
+Add an anonymous variable described by the keyword arguments `kwargs` that
+belongs to the model `m`. Note that if the `basename` argument is specified
+to `varname`, the variable created is not anonymous but has the name `varname`.
+
+    @variable(m, expr, args...; kwargs...)
+
+Add a variable described by the expression `expr`, the positional arguments
+`args` and the keyword arguments `kwargs` that belongs to the model `m`.
+The expression `expr` can either be (note that in the following the symbol
+`<=` can be used instead of `≤` and the symbol `>=`can be used instead of
+`≥`)
+
+* of the form `varexpr` creating variables described by `varexpr`;
+* of the form `varexpr ≤ ub` (resp. `varexpr ≥ lb`) creating variables described by
+  `varexpr` with upper bounds given by `ub` (resp. lower bounds given by `lb`);
+* of the form `varexpr == value` creatin variables described by `varexpr` with
+  fixed values given by `value`;
+* of the form `lb ≤ varexpr ≤ ub` or `ub ≥ varexpr ≥ lb` creating variables
+  described by `varexpr` with lower bounds given by `lb` and upper bounds given
+  by `ub`.
+
+The expression `varexpr` can either be
+
+* of the form `varname` creating a scalar real variable of name `varname`;
+* of the form `varname[axes]` creating an `Array` or `JuMPArray` of scalar
+  real variables of names `varname[...]` for each indices `...` of the axes
+  `axes`;
+* of the form `[axes]` creating an `Array` or `JuMPArray` of anonymous scalar
+  real variables. Note that if the `basename` argument is specified to
+  `varname`, the variables created are note anonymous but have the names
+  `varname[...]` for each indices `...` of the axes `axes` as in the previous
+  case.
+
+The expression `axes` can either be
+
+* of the form `1:s1,1:s2,...,1:sn` creating an `n`-dimensional `Array` of size
```

Comment:
This misleadingly implies that you can't mix the different axis types.

Reply:
What do you mean ? All the axis have type `Base.OneTo` here, haven't they ?

Reply:
I mean, it's easy to infer from this list that you shouldn't write `[i=1:N, foo_set]`.